File: src/rust/planning_poker/src/usecases/join_user.rs
use gloo::console::console_dbg;

use crate::{
    domains::{
        event::DomainEventKind,
        game_player::GamePlayerId,
        invitation::InvitationSignature,
        join_service::{HaveJoinService, JoinService},
        user::{HaveUserRepository, UserId, UserRepository},
    },
    infrastructures::event::event_dispatcher::{EventDispatcher, HaveEventDispatcher},
    utils::types::LocalBoxFuture,
};

pub trait JoinUserDependency: HaveUserRepository + HaveEventDispatcher + HaveJoinService {}

#[derive(Clone, Debug, PartialEq)]
pub enum JoinUserOutput {
    NotFound,
    JoinFailed,
}

pub trait JoinUser {
    fn execute(
        &self,
        signature: InvitationSignature,
        user_id: UserId,
    ) -> LocalBoxFuture<'_, Result<GamePlayerId, JoinUserOutput>>;
}

impl<T: JoinUserDependency> JoinUser for T {
    fn execute(
        &self,
        signature: InvitationSignature,
        user_id: UserId,
    ) -> LocalBoxFuture<'_, Result<GamePlayerId, JoinUserOutput>> {
        let repository = self.get_user_repository();
        let dispatcher = self.get_event_dispatcher();
        let join_service = self.get_join_service();

        let fut = async move {
            let user = repository.find_by(user_id).await;

            if user.is_none() {
                return Err(JoinUserOutput::NotFound);
            }
            let user = user.unwrap();

            let event = join_service.join(&user, signature).await;
            if let Some(DomainEventKind::UserInvited { game_player_id, .. }) = event {
                dispatcher.dispatch(&event.unwrap());
                Ok(game_player_id)
            } else {
                Err(JoinUserOutput::JoinFailed)
            }
        };

        Box::pin(fut)
    }
}

File: app/src/usecases/join-user.ts
import { EventDispatcher, UseCase } from "./base";
import { GameRepository } from "@/domains/game-repository";
import * as Invitation from "@/domains/invitation";
import * as User from "@/domains/user";
import * as Game from "@/domains/game";
import { UserRepository } from "@/domains/user-repository";

export interface JoinUserUseCaseInput {
  signature: Invitation.T;
  userId: User.Id;
}

export type JoinUserUseCaseOutput =
  | { kind: "success"; game: Game.T }
  | { kind: "notFoundUser" }
  | { kind: "notFoundGame" }
  | { kind: "joinFailed" };

export class JoinUserUseCase implements UseCase<JoinUserUseCaseInput, Promise<JoinUserUseCaseOutput>> {
  constructor(
    private dispatcher: EventDispatcher,
    private userRepository: UserRepository,
    private gameRepository: GameRepository
  ) {}

  async execute(input: JoinUserUseCaseInput): Promise<JoinUserUseCaseOutput> {
    const user = await this.userRepository.findBy(input.userId);

    if (!user) {
      return { kind: "notFoundUser" };
    }
    const game = await this.gameRepository.findByInvitation(input.signature);
    if (!game) {
      return { kind: "notFoundGame" };
    }

    try {
      const [newGame, event] = Game.joinUserAsPlayer(game, user.id, input.signature);
      await this.gameRepository.save(newGame);

      this.dispatcher.dispatch(event);

      return { kind: "success", game: newGame };
    } catch (e) {
      console.error(e);

      return { kind: "joinFailed" };
    }
  }
}
